node/src/neighborhood/overall_connection_status.rs

```diff
@@ -814,147 +771,222 @@ mod tests {
-        let mut subject = OverallConnectionStatus::new(vec![initial_node_descriptor]);
-        let (recipient, recording_arc) = make_node_to_ui_recipient();
-        subject.stage = OverallConnectionStage::ThreeHopsRouteFound;
-        let system = System::new("test");
+        let initial_stage = OverallConnectionStage::ThreeHopsRouteFound;
+        let event = ConnectionProgressEvent::StandardGossipReceived;
+        let can_make_routes = false;
+
+        let (stage, message_opt) =
+            result_when_neighborship_is_established_and_can_make_routes_is_updated(
+                initial_stage,
+                event,
+                can_make_routes,
+                "doesn_t_send_a_message_to_ui_in_case_connection_drops_from_three_hops_to_connected_to_neighbor"
+            );
+
+        assert_eq!(stage, OverallConnectionStage::ConnectedToNeighbor);
+        assert_eq!(message_opt, None);
+    }
+
+    #[test]
+    fn progress_done_by_one_connection_progress_can_not_be_overridden_by_other_in_overall_connection_progress(
+    ) {
+        let ip_addr_1 = make_ip(1);
+        let ip_addr_2 = make_ip(2);
+        let mut subject = OverallConnectionStatus::new(vec![
+            make_node_descriptor(ip_addr_1),
+            make_node_descriptor(ip_addr_2),
+        ]);
+        let (node_to_ui_recipient, _) = make_node_to_ui_recipient();
         subject.update_connection_stage(
-            ip_addr,
+            ip_addr_1,
             ConnectionProgressEvent::TcpConnectionSuccessful,
-            &recipient,
+            &node_to_ui_recipient,
+        );
+        subject.update_connection_stage(
+            ip_addr_1,
+            ConnectionProgressEvent::IntroductionGossipReceived(make_ip(3)),
+            &node_to_ui_recipient,
+        );
+        subject.update_connection_stage(
+            ip_addr_2,
+            ConnectionProgressEvent::TcpConnectionSuccessful,
+            &node_to_ui_recipient,
         );
-        subject.update_can_make_routes(false);
 
         subject.update_connection_stage(
-            ip_addr,
-            ConnectionProgressEvent::StandardGossipReceived,
-            &recipient,
+            ip_addr_2,
+            ConnectionProgressEvent::PassGossipReceived(make_ip(4)),
+            &node_to_ui_recipient,
         );
 
-        System::current().stop();
-        assert_eq!(system.run(), 0);
-        let recording = recording_arc.lock().unwrap();
-        assert_eq!(recording.len(), 0);
+        assert_eq!(subject.stage, OverallConnectionStage::ConnectedToNeighbor);
     }
 
-    fn make_node_descriptor_from_ip(ip_addr: IpAddr) -> NodeDescriptor {
-        NodeDescriptor {
-            blockchain: Chain::EthRopsten,
-            encryption_public_key: PublicKey::from(vec![0, 0, 0]),
-            node_addr_opt: Some(NodeAddr::new(&ip_addr, &vec![1, 2, 3])),
+    fn result_when_neighborship_is_established_and_can_make_routes_is_updated(
```

Comment:
Consider this name: 
`fn stage_and_ui_message_by_connection_progress_event _and_can_make_routes(...){...}`